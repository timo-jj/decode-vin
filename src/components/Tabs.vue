<template>
  <article class="tabs">
    <section class="tabs__controls">
      <button
        v-for="tab in state.tabs"
        :key="tab.id"
        :class="[
          'button',
          {
            'button--active': state.activeTab === tab.id,
          },
        ]"
        @click="setActiveTab(tab.id)"
      >
        {{ tab.title }}
      </button>
    </section>
  </article>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import { Tabs } from '@/@types/ux'

export default defineComponent({
  name: 'Tabs',

  emits: ['select-active-tab'],

  data() {
    return {
      state: {
        tabs: [
          {
            id: 0,
            title: 'Single Vin',
            flag: 'single-vin',
          },
          {
            id: 1,
            title: 'Bulk Vin',
            flag: 'bulk-vin',
          },
          {
            id: 2,
            title: 'Manual Entry',
            flag: 'manual-entry',
          },
        ] as Tabs[],
        activeTab: 0 as number,
      },
    }
  },

  methods: {
    setActiveTab(tabID: number): void {
      this.state.activeTab = tabID
      this.$emit('select-active-tab', this.state.activeTab)
    },
  },
})
</script>

<!-- script setup syntax
<script setup lang="ts">
import { defineEmits, reactive, defineExpose } from 'vue'

import { Tabs } from '@/@types/ux'

const emit = defineEmits<{
  (eventName: 'select-active-tab', activeTab: number): void
}>()

const state = reactive({
  tabs: [
    {
      id: 0,
      title: 'Single Vin',
      flag: 'single-vin',
    },
    {
      id: 1,
      title: 'Bulk Vin',
      flag: 'bulk-vin',
    },
    {
      id: 2,
      title: 'Manual Entry',
      flag: 'manual-entry',
    },
  ] as Tabs[],
  activeTab: 0 as number,
})

const setActiveTab = (tabID: number): void => {
  state.activeTab = tabID
  emit('select-active-tab', state.activeTab)
}
</script>
-->

<style lang="scss" scoped>
.tabs {
  &__controls {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    height: 80px;
    background: #ffffff;
    border-bottom: 1px solid #a3a3a3;

    .button {
      border: 0;
      outline: 0;
      text-transform: uppercase;
      background: #ffffff;
      transition: all 0.2s ease-in-out;
      font-size: 16px;
      color: #a3a3a3;
      height: 100%;
      padding: 0;

      &--active {
        box-shadow: inset 0px -10px 0px -3px #1876d2;
        color: #1876d2;
      }
    }
  }
}
</style>
