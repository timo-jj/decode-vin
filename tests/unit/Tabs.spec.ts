import { shallowMount } from '@vue/test-utils'

import Tabs from '@/components/Tabs.vue'

const wrapper = shallowMount(Tabs, {
  emits: ['select-active-tab'],
} as any)

describe('Tabs', () => {
  describe('when rendering', () => {
    it('it renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('on tab click', () => {
    it('should emit the event', async () => {
      await wrapper.find('button').trigger('click')

      expect(wrapper.emitted('select-active-tab')).toBeTruthy()
    })
  })
})
