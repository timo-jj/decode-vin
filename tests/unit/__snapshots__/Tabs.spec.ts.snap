// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs when rendering it renders correctly 1`] = `
<article
  class="tabs"
>
  <section
    class="tabs__controls"
  >
    
    <button
      class="button button--active"
    >
      Single Vin
    </button>
    <button
      class="button"
    >
      Bulk Vin
    </button>
    <button
      class="button"
    >
      Manual Entry
    </button>
    
  </section>
</article>
`;
